Adding first-class support for versioned store

brewkode opened the pull request:
Added a versioned store that wraps around any suuchi.Store implementation and provides versioning support for keys
Differentiates Version records & Actual data records with a prefix so that scans for versions (if need be) can be sped up
Explicitly made this project source/target to be java 1.8

@ashwanthkumar Please take a look.

brewkode committed:
Adding first-class support for versioned store

Added a versioned store that wraps around any suuchi.Store implementation and provides versioning support for keys
Differentiates Version records & Actual data records with a prefix so that scans for versions (if need be) can be sped up
Explicitly made this project source/target to be java 1.8

brewkode committed:
Rocks store supports remove as well

ashwanthkumar commented on suuchi-core/src/main/scala/in/ashwanthkumar/suuchi/store/Store.scala:
```diff
@@ -8,4 +8,8 @@ trait WriteStore {
 }
 
-trait Store extends ReadStore with WriteStore
+trait DeleteStore {
```
@brewkode Should we push this `remove` as part of `WritableStore`? Isn't it safe to assume that WriteStores also supports Deletes?

ashwanthkumar commented on suuchi-core/src/main/scala/in/ashwanthkumar/suuchi/store/VersionedStore.scala:
```diff
@@ -0,0 +1,87 @@
+}
+
+object VersionedStore {
+  val VERSION_PREFIX = "V_"
```
@brewkode Should we convert these `*_PREFIX` to `byte[]` of fixed length instead of a string?

brewkode replied:
Done.

ashwanthkumar commented on suuchi-core/src/main/scala/in/ashwanthkumar/suuchi/store/VersionedStore.scala:
```diff
@@ -0,0 +1,87 @@
+
+class VersionedStore(store: Store, numVersions: Int) extends Store with DeleteStore with DateUtils {
+  import VersionedStore._
+  val SYNC_SLOTS = Array.fill(8192)(new Object)
```
@brewkode Can we please make this `8192` as a default value but allow it to be overridden via constructor?

brewkode replied:
Done.

brewkode committed:
Fixing weird way to specify javacArg in scala-maven-plugin

ashwanthkumar commented on suuchi-core/src/main/scala/in/ashwanthkumar/suuchi/store/VersionedStore.scala:
```diff
@@ -0,0 +1,87 @@
+  def remove(key: Array[Byte]) = {
+    val versions = getVersions(key)
+    removeVersion(key)
+    versions.foreach(v => removeData(key, v))
```
Can we convert this to a `forall` so we could automatically return a boolean given `removeData` also returns a boolean value?

brewkode replied:
Done.

ashwanthkumar commented on suuchi-core/src/main/scala/in/ashwanthkumar/suuchi/store/VersionedStore.scala:
```diff
@@ -0,0 +1,87 @@
+    store.get(dkey(key, version))
+  }
+
+  override def put(key: Array[Byte], value: Array[Byte]): Boolean = {
```
Should we also support a `put` variant with `version` argument given that's what we'll be using this with?

brewkode replied:
I thought about it, but, having a `put`/`remove` variant with a `version` argument means, it'd go into the `Store` contract too, which is unnecessary at this point.

I'd rather have a wrapper store, to support that specific use case and deal with it separately.

ashwanthkumar replied:
Sounds valid. In which case we might have to wrap VersionedStore as an abstract class or something and support a method which returns a version based on `(key, value)`.

brewkode replied:
I guess, I got side tracked here. I'll roll out some changes to this shortly.

ashwanthkumar commented on suuchi-core/src/main/scala/in/ashwanthkumar/suuchi/store/VersionedStore.scala:
```diff
@@ -0,0 +1,87 @@
+    val monitor = SYNC_SLOTS(absHash % SYNC_SLOTS.length)
+    val versions  = monitor.synchronized {
+      val vRecord = store.get(versionKey)
+      val updatedVersions = vRecord.map(bytes => version :: Versions.fromString(new String(bytes)).toList).getOrElse(List(version))
```
@brewkode Here the `updatedVersions` seems to assume the last values and no other ordering, while it's very natural to assume these to be that of naturally ordered versions. Can we make that explicit? Probably support versionOrdering as a constructor argument or something similar?

brewkode replied:
If the `version` is a `long`, I don't see why you need to specify a different ordering. Now, we have only 2 possibilities...
- You want to use a different ordering for `long` itself.. or,
- You are encoding something as a `long` and want the system to use it as `version`. The catch though is, we are not exposing the `version` information to be meddled with. We know HBase does it, but, we also know, what kind of problems it has because of unintended user behavior. So, I would rather not expose what `version` is outside and manage it internally. However from an internal cleanliness perspective, we could have a `Version` type and provide ordering with it.

ashwanthkumar replied:
That makes sense, but my question was more around the ordering when we do a `.take(numVersions)` in the next line. We don't sort, so the ordering that's maintained here is the insertion order than the natural ordering of `long` type. Which was a little confusing for me and might cause issues with out of order updates.

ashwanthkumar replied:
May be sometimes we need insertion ordering like a CQRS or an event log for example, and using `now` would  make perfect sense and also make `sort` redundant. In all other cases we might have to do `sort` on the versions based on an `Ordering` - hence I feel we need an ordering which can be implicit but that gives developers greater control.

brewkode replied:
I had ordering before, removed because I thought, inserts are managed by us, so there was no need to have ordering explicitly. I get what you're saying. Next cut should address this.

ashwanthkumar commented on suuchi-core/src/main/scala/in/ashwanthkumar/suuchi/store/VersionedStore.scala:
```diff
@@ -0,0 +1,87 @@
+    val monitor = SYNC_SLOTS(absHash % SYNC_SLOTS.length)
+    val versions  = monitor.synchronized {
+      val vRecord = store.get(versionKey)
+      val updatedVersions = vRecord.map(bytes => version :: Versions.fromString(new String(bytes)).toList).getOrElse(List(version))
```
Also wondering if we need to do SerDe of the versions given we're not doing anything with them.

brewkode replied:
Dint quite understand it.

ashwanthkumar replied:
In this line we do `Versions.fromString(new String(bytes)).toList` and on the next we do `Versions.toString(updatedVersions.take(numVersions)).getBytes`. We de-serialise and serialise the versions but don't do anything with them. I was wondering if we can avoid it completely.

brewkode replied:
we actually add an item to the head of the list and take only the first 3. Perhaps, that's not clear. I pushed a minor code change in an attempt to make it explicit.

ashwanthkumar commented on suuchi-core/src/main/scala/in/ashwanthkumar/suuchi/store/VersionedStore.scala:
```diff
@@ -0,0 +1,87 @@
+
+  def atomicUpdate(key: Array[Byte], version: Long) = {
+    val versionKey = vkey(key)
+    val absHash = math.abs(MurmurHash3.arrayHash(versionKey))
```
@brewkode Minor thing, I read that `MurmurHash3.bytesHash` is a little faster than `MurmurHash3.arrayHash`. 

Ref 
- https://github.com/scala/scala/blob/v2.11.8/src/library/scala/util/hashing/MurmurHash3.scala#L128-L130
- I've done this change as part of #43 as well for CHRing.

brewkode replied:
Done.

ashwanthkumar commented on suuchi-core/src/main/scala/in/ashwanthkumar/suuchi/store/VersionedStore.scala:
```diff
@@ -0,0 +1,87 @@
+  def remove(key: Array[Byte]) = {
+    val versions = getVersions(key)
+    removeVersion(key)
+    versions.foreach(v => removeData(key, v))
```
Similar to `put`discussion, should we also allow a `remove` with `version` semantics. It's starting to feel more like HBase Client API now 😨

brewkode replied:
See my comments on exposing version.

ashwanthkumar commented on suuchi-rocksdb/src/main/scala/in/ashwanthkumar/suuchi/store/rocksdb/RocksDbStore.scala:
```diff
@@ -27,4 +27,9 @@ class RocksDbStore(config: RocksDbConfiguration) extends Store with Logging {
   }
+
+  override def remove(key: Array[Byte]): Boolean = {
+    db.remove(key)
```
@brewkode Should we wrap this db.remove with `Try( ... ).isSuccess` like in `put` ?

ashwanthkumar replied:
Little off the record - Above the changes I see the `put` and `get` are synchronized in RocksStore. We don't need it right?

ashwanthkumar commented on suuchi-core/src/main/scala/in/ashwanthkumar/suuchi/store/VersionedStore.scala:
```diff
@@ -0,0 +1,87 @@
+    // atomically update version metadata
+    val versions = atomicUpdate(key, nowMs)
+
+    // remove oldest version, if we've exceeded max # of versions per record
```
Might be too much and I know you're going to scold me for this, but still 😛  - do you think it's worthwhile to off-load this task on a thread and forget about it? That way the caller is not blocked for the purge to happen.

brewkode replied:
Not worth it, at this point.

ashwanthkumar replied:
👍

brewkode committed:
incorporating #44 review comments

1. Removed DeleteStore trait.
2. used bytesHash
3. data & version prefixes are now fixed size bytes

brewkode committed:
Wrapping remove with a Try

brewkode committed:
Minor cleanup to use toBytes instead of toString

ashwanthkumar commented:
Ok just now saw the change. I saw that we're adding it to the head earlier too. The ordering maintained here is the insertion order la ji. 

I think its not obvious because the version is the timestamp now, one of my earlier comments was make that configurable such that given a key and a value a method has to return a long. And we need that for our use case la. In that scenario to handle out of order updates we need to sort the versions. Appo we will need ordering la. 

Text by Ashwanth, typos by Lumia

-----Original Message-----
From: "Sriram" notifications@github.com
Sent: ‎07-‎10-‎2016 14:49
To: "ashwanthkumar/suuchi" suuchi@noreply.github.com
Cc: "Ashwanth Kumar" ashwanthkumar@googlemail.com; "Mention" mention@noreply.github.com
Subject: Re: [ashwanthkumar/suuchi] Adding first-class support for versionedstore (#44)

@brewkode commented on this pull request.

In suuchi-core/src/main/scala/in/ashwanthkumar/suuchi/store/VersionedStore.scala:

> -    // remove oldest version, if we've exceeded max # of versions per record
> -    if(versions.size > numVersions) removeData(key, versions.min)
>   +
> -    // Write out the actual data record
> -    store.put(dkey(key, versions.max), value)
>   +
> -  }
>   +
> -  def atomicUpdate(key: Array[Byte], version: Long) = {
> -    val versionKey = vkey(key)
> -    val absHash = math.abs(MurmurHash3.arrayHash(versionKey))
> -    // Synchronizing the version metadata update part alone
> -    val monitor = SYNC_SLOTS(absHash % SYNC_SLOTS.length)
> -    val versions  = monitor.synchronized {
> -      val vRecord = store.get(versionKey)
> -      val updatedVersions = vRecord.map(bytes => version :: Versions.fromString(new String(bytes)).toList).getOrElse(List(version))
>   we actually add an item to the head of the list and take only the first 3. Perhaps, that's not clear. I pushed a minor code change in an attempt to make it explicit.
>   —
>   You are receiving this because you were mentioned.
>   Reply to this email directly, view it on GitHub, or mute the thread.

ashwanthkumar commented:
Ok ignore my previous comment. Wrong context.

brewkode committed:
Extracting out ordering + version identification from msg

1. VersionedBy provides a way to define your own versioning scheme.
2. ByWriteTimestamp is a VersionedBy implementation based on current write timestamp
3. Changed the way the list of versions are serialized and stored in the underlying store - using proper primitive serialization mechanism instead of toString.getBytes
4. Version can only be Long. Dint want to compromise on it, as serialization becomes unnecessarily complicated

brewkode committed:
Added spec for Primitives serde + fixed a state issue with test

ashwanthkumar commented on suuchi-core/src/main/scala/in/ashwanthkumar/suuchi/store/PrimitivesSerDeUtils.scala:
```diff
@@ -0,0 +1,13 @@
+  * */
+  def longToBytes(instance: Long) = ByteBuffer.allocate(8).putLong(instance).array()
+  def intToBytes(instance: Int) = ByteBuffer.allocate(4).putInt(instance).array()
+  def bytesToInt(bytes: Array[Byte]) = ByteBuffer.wrap(bytes).getLong
```
Should be `getInt` no?

ashwanthkumar replied:
Looks like this is fixed in a later commit, please ignore.

ashwanthkumar commented on suuchi-core/src/main/scala/in/ashwanthkumar/suuchi/store/VersionedStore.scala:
```diff
@@ -5,20 +5,38 @@ import scala.util.hashing.MurmurHash3
+  }
+
+  def toBytes(versions: List[Long]): Array[Byte] = {
+    val numVersions = versions.size.toByte
```
Using a `byte` limits the number of versions we support isn't? That's intentional? Int is long enough generally but `Byte` feels so small. Thoughts?

brewkode replied:
Are you saying, you would need more than 256 versions per record? I don't see this as limiting at all.

ashwanthkumar replied:
Okay for now I guess we can stick with that and later may be change that.

ashwanthkumar committed:
Adding simple test for Versions's Long SerDe

ashwanthkumar committed:
Fixing a bug where we were always inserting the current data as max(all known versions for this key) instead it should be using the currentVersion to write it. Added a failing test for the same

ashwanthkumar commented:
@brewkode I looked into this branch and make a couple of small changes, rest LGTM. You can merge this anytime.

ashwanthkumar committed:
Trimming down extra empty lines

ashwanthkumar committed:
making the spec name more meaningful

ashwanthkumar committed:
Logging exceptions in RocksDB store so we can debug when something actually goes wrong

ashwanthkumar commented:
@brewkode Merging this once the build goes green.
